### PR TITLE
feat(streaming): handle watermark for mv on mv

### PR DESCRIPTION
**Pull request by chenzl25**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

- Ignore watermark during snapshot consumption and forward watermark messages to downstream after that.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features).
- [x] I have demonstrated that backward compatibility is not broken by breaking changes and created issues to track deprecated features to be removed in the future. (Please refer the issue)
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

If your pull request contains user-facing changes, please specify the types of the changes, and create a release note. Otherwise, please feel free to remove this section.

### Types of user-facing changes

Please keep the types that apply to your changes, and remove those that do not apply.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

Please create a release note for your changes. In the release note, focus on the impact on users, and mention the environment or conditions where the impact may occur.

### CODE REVIEW

File: src/stream/src/executor/backfill.rs [hunk 1]
```diff
@@ -381,7 +381,12 @@ where
 
     fn mapping_message(msg: Message, upstream_indices: &[usize]) -> Message {
         match msg {
-            Message::Barrier(_) | Message::Watermark(_) => msg,
+            Message::Barrier(_) => msg,
+            Message::Watermark(watermark) => Message::Watermark(Watermark {
+                col_idx: upstream_indices[watermark.col_idx],
```
Comment (BugenZhao) [inline src/stream/src/executor/backfill.rs:386]: The `upstream_indices` seems to be a projection, that is, we should find the position of the `col_idx` value and ignore the message if not exists.

File: src/stream/src/executor/backfill.rs [hunk 2]
```diff
@@ -381,7 +381,12 @@ where
...
+                data_type: watermark.data_type,
+                val: watermark.val,
```
Comment (BugenZhao) [inline src/stream/src/executor/backfill.rs:388]:
```suggestion
                ..watermark
```

File: src/stream/src/executor/backfill.rs (hunk 1 above)
Comment (chenzl25) [reply src/stream/src/executor/backfill.rs:386]: Oops! what a mistake.

File: src/stream/src/executor/rearranged_chain.rs [hunk 3]
```diff
@@ -135,10 +141,11 @@ impl RearrangedChainExecutor {
     async fn execute_inner(mut self) {
         // 0. Project the upstream with `upstream_indices`.
         let upstream_indices = self.upstream_indices.clone();
-        let mut upstream = self
-            .upstream
-            .execute()
-            .map(move |result| result.map(|msg| mapping(&upstream_indices, msg)));
```
Comment (BugenZhao) [inline src/stream/src/executor/rearranged_chain.rs:141]: We should be able to use `.map(..).flatten()` here.

**Review by BugenZhao — APPROVED** [review]

File: src/stream/src/executor/rearranged_chain.rs (hunk 3 above)
Comment (chenzl25) [reply src/stream/src/executor/rearranged_chain.rs:141]: Could you please elaborate a bit more？

Comment (BugenZhao) [reply src/stream/src/executor/rearranged_chain.rs:141]:
```rust
        let mut upstream = self
            .upstream
            .execute()
            .map(move |result| result.map(|msg| mapping(&upstream_indices, msg)))
            .flatten();
```

Comment (chenzl25) [reply src/stream/src/executor/rearranged_chain.rs:141]: But `mapping` returns an Option, which seems unable to be flattened.

**Comment by st1page** [comment]
A random thought, if a watermark column exists as the first column in the pk, can we emit the watermark when backfill? Even maybe we can do some optimization for backfill with that watermark?

**Comment by chenzl25** [comment]
Good catch. I think if the first column of the pk is ascending, we can emit watermark once it passes the first column of  `current_pos` of the backfill progress, otherwise we can emit the first column of  `current_pos` itself.

**Pull request merged**